File: app/src/main/java/za/co/woolworths/financial/services/android/ui/fragments/contact_us/ContactUsFinancialServiceFragment.java
package za.co.woolworths.financial.services.android.ui.fragments.contact_us;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.awfs.coordination.R;

import java.util.List;

import za.co.woolworths.financial.services.android.ui.activities.CustomPopUpWindow;
import za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigator;
import za.co.woolworths.financial.services.android.util.Utils;

/**
 * Created by W7099877 on 01/12/2016.
 */

public class ContactUsFinancialServiceFragment extends Fragment implements View.OnClickListener {
	private static final int REQUEST_CALL = 1;
	Intent callIntent;

	private BottomNavigator mBottomNavigator;

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.contact_us_financial_services, container, false);
		view.findViewById(R.id.localCaller).setOnClickListener(this);
		view.findViewById(R.id.internationalCaller).setOnClickListener(this);
		view.findViewById(R.id.blackCrediCardQuery).setOnClickListener(this);
		view.findViewById(R.id.complains).setOnClickListener(this);
		view.findViewById(R.id.creditCardQuery).setOnClickListener(this);
		view.findViewById(R.id.wRewardsQuery).setOnClickListener(this);
		view.findViewById(R.id.insuranceClaim).setOnClickListener(this);
		view.findViewById(R.id.paymentQuery).setOnClickListener(this);
		view.findViewById(R.id.storeCardPesonalLoanQuery).setOnClickListener(this);
		view.findViewById(R.id.proofOfIncome).setOnClickListener(this);
		view.findViewById(R.id.technical).setOnClickListener(this);

		try {
			mBottomNavigator = (BottomNavigator) getActivity();
			mBottomNavigator.setTitle(getActivity().getResources().getString(R.string.contact_us_financial_services));
		} catch (ClassCastException ex) {
			Log.e("Interface", ex.toString());
		}

		return view;
	}

	@Override
	public void onClick(View v) {
		switch (v.getId()) {
			case R.id.localCaller:
				Utils.makeCall(getActivity(), getActivity().getResources().getString(R.string.fs_local_caller_number));
				break;
			case R.id.internationalCaller:
				Utils.makeCall(getActivity(), getActivity().getResources().getString(R.string.fs_inter_national_caller_number));
				break;
			case R.id.blackCrediCardQuery:
				sendEmail(getActivity().getResources().getString(R.string.email_black_credit_card_query), getActivity().getResources().getString(R.string.txt_black_credit_card_query));
				break;
			case R.id.complains:
				sendEmail(getActivity().getResources().getString(R.string.email_complaints), getActivity().getResources().getString(R.string.txt_complaint));
				break;
			case R.id.creditCardQuery:
				sendEmail(getActivity().getResources().getString(R.string.email_credicard_query), getActivity().getResources().getString(R.string.txt_cc_query));
				break;
			case R.id.wRewardsQuery:
				sendEmail(getActivity().getResources().getString(R.string.email_wrewards_query), getActivity().getResources().getString(R.string.txt_wrewards_query));
				break;
			case R.id.insuranceClaim:
				sendEmail(getActivity().getResources().getString(R.string.email_insurance_claim), getActivity().getResources().getString(R.string.txt_insurance_claim));
				break;
			case R.id.paymentQuery:
				sendEmail(getActivity().getResources().getString(R.string.email_payment_query), getActivity().getResources().getString(R.string.txt_payment_query));
				break;
			case R.id.storeCardPesonalLoanQuery:
				sendEmail(getActivity().getResources().getString(R.string.email_sc_and_pl_query), getActivity().getResources().getString(R.string.txt_sc_and_pl_query));
				break;
			case R.id.proofOfIncome:
				sendEmail(getActivity().getResources().getString(R.string.email_proof_of_income), getActivity().getResources().getString(R.string.txt_proof_of_income));
				break;
			case R.id.technical:
				sendEmail(getActivity().getResources().getString(R.string.email_technical), getActivity().getResources().getString(R.string.txt_technical_problem));
				break;
		}
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
		switch (requestCode) {
			case REQUEST_CALL:
				if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
					startActivity(callIntent);
				}
		}
	}

	public void sendEmail(String emailId, String subject) {
		Intent emailIntent = new Intent(Intent.ACTION_SENDTO);
		emailIntent.setData(Uri.parse("mailto:" + emailId +
				"?subject=" + Uri.encode(subject) +
				"&body=" + Uri.encode("")));

		PackageManager pm = getActivity().getPackageManager();
		List<ResolveInfo> listOfEmail = pm.queryIntentActivities(emailIntent, 0);
		if (listOfEmail.size() > 0) {
			startActivity(emailIntent);
		} else {
			Utils.displayValidationMessage(getActivity(),
					CustomPopUpWindow.MODAL_LAYOUT.INFO,
					getActivity().getResources().getString(R.string.contact_us_no_email_error)
							.replace("email_address", emailId).replace("subject_line", subject));
		}
	}

	@Override
	public void onHiddenChanged(boolean hidden) {
		super.onHiddenChanged(hidden);
		if (!hidden) {
			if (mBottomNavigator != null) {
				mBottomNavigator.setTitle(getActivity().getResources().getString(R.string.contact_us_financial_services));
				mBottomNavigator.displayToolbar();
				mBottomNavigator.showBackNavigationIcon(true);
			}
		}
	}
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/fragments/contact_us/ContactUsFinancialServiceFragment.kt
package za.co.woolworths.financial.services.android.ui.fragments.contact_us

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.awfs.coordination.R
import com.awfs.coordination.databinding.ContactUsFinancialServicesBinding
import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties
import za.co.woolworths.financial.services.android.ui.activities.account.MyAccountActivity
import za.co.woolworths.financial.services.android.ui.activities.account.sign_in.whatsapp.WhatsAppChatToUs
import za.co.woolworths.financial.services.android.ui.activities.account.sign_in.whatsapp.WhatsAppChatToUs.Companion.CONTACT_US
import za.co.woolworths.financial.services.android.ui.activities.account.sign_in.whatsapp.WhatsAppChatToUs.Companion.FEATURE_WHATSAPP
import za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigationActivity
import za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigator
import za.co.woolworths.financial.services.android.ui.extension.bindString
import za.co.woolworths.financial.services.android.ui.fragments.contact_us.enquiry.list.EnquiriesListFragment
import za.co.woolworths.financial.services.android.ui.views.actionsheet.WhatsAppUnavailableFragment
import za.co.woolworths.financial.services.android.util.ScreenManager
import za.co.woolworths.financial.services.android.util.Utils

class ContactUsFinancialServiceFragment : Fragment(R.layout.contact_us_financial_services), View.OnClickListener {

    private lateinit var binding: ContactUsFinancialServicesBinding
    private val contactUsModel = ContactUsModel()

    companion object {
        private const val REQUEST_CALL = 1

    }

    private var callIntent: Intent? = null
    private var mBottomNavigator: BottomNavigator? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is BottomNavigationActivity)
            mBottomNavigator = context
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding = ContactUsFinancialServicesBinding.bind(view)

        binding.apply {
            setupToolbar()
            with(contactUsModel) {
                contactUsFinancialServicesEmail()?.apply {
                    val localCallerRow = layoutInflater.inflate(
                        R.layout.contact_us_email_item,
                        contactFinancialServicesEmailLinearLayout,
                        false
                    )
                    val contactUsEmailTextView =
                        localCallerRow.findViewById<TextView>(R.id.contactUsEmailTextView)
                    val contactUsEmailDescriptionTextView =
                        localCallerRow.findViewById<TextView>(R.id.contactUsEmailDescriptionTextView)
                    val contactUsEmailDescriptionDivider =
                        localCallerRow.findViewById<View>(R.id.contactUsEmailDescriptionDivider)

                    contactUsEmailTextView?.text = bindString(R.string.send_enquiry)
                    contactUsEmailDescriptionTextView?.visibility = GONE
                    contactUsEmailDescriptionDivider?.visibility = GONE
                    localCallerRow?.setOnClickListener {
                        if (activity is BottomNavigationActivity)
                            mBottomNavigator?.pushFragmentSlideUp(EnquiriesListFragment())
                        else
                            (activity as? MyAccountActivity)?.replaceFragment(EnquiriesListFragment())
                    }

                    contactFinancialServicesEmailLinearLayout?.addView(localCallerRow)

                }

                contactUsFinancialServicesCall()?.apply {
                    options?.forEachIndexed { index, item ->
                        val localCallerRow = layoutInflater.inflate(
                            R.layout.contact_us_call_options_item,
                            callUsLinearLayoutContainer,
                            false
                        )
                        val localCallerTextView =
                            localCallerRow.findViewById<TextView>(R.id.localCallerTextView)
                        val localCallerPhoneNumberTextView =
                            localCallerRow.findViewById<TextView>(R.id.localCallerPhoneNumberTextView)

                        localCallerTextView?.text = item.key
                        localCallerPhoneNumberTextView?.text = item.value
                        localCallerRow?.tag = index
                        localCallerRow?.setOnClickListener { Utils.makeCall(item.value) }

                        callUsLinearLayoutContainer?.addView(localCallerRow)
                    }

                    operationHoursTextView?.text = operatingHours
                }
            }

            showWhatsAppChatWithUs()
            contactUsChatToUsRelativeLayout?.setOnClickListener(this@ContactUsFinancialServiceFragment)
        }
    }

    private fun ContactUsFinancialServicesBinding.showWhatsAppChatWithUs() {
        with(WhatsAppChatToUs()) {
            if (isChatWithUsEnabledForContactUs) {
                chatWithUsLinearLayout?.visibility = VISIBLE
                if (isCustomerServiceAvailable) {
                    whatsAppIconImageView?.setImageResource(R.drawable.icon_whatsapp_grey)
                    whatsAppTitleTextView?.setTextColor(Color.BLACK)
                    whatsAppNextIconImageView?.alpha = 1f
                } else {
                    whatsAppIconImageView?.setImageResource(R.drawable.icon_whatsapp_grey)
                    whatsAppNextIconImageView?.alpha = 0.4f
                    activity?.let { activity -> whatsAppTitleTextView?.setTextColor(ContextCompat.getColor(activity, R.color.unavailable)) }
                }
            } else {
                chatWithUsLinearLayout?.visibility = GONE
            }
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.contactUsChatToUsRelativeLayout -> {
                if (!WhatsAppChatToUs().isCustomerServiceAvailable) {
                    val whatsAppUnavailableFragment = WhatsAppUnavailableFragment()
                    activity?.supportFragmentManager?.let { supportFragmentManager -> whatsAppUnavailableFragment.show(supportFragmentManager, WhatsAppUnavailableFragment::class.java.simpleName) }
                    return
                }
                activity?.apply { Utils.triggerFireBaseEvents(FirebaseManagerAnalyticsProperties.WHATSAPP_CONTACT_US, this) }
                ScreenManager.presentWhatsAppChatToUsActivity(activity, FEATURE_WHATSAPP, CONTACT_US)
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            REQUEST_CALL -> if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                activity?.startActivity(callIntent)
            }
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden)
            setupToolbar()
    }

    private fun setupToolbar() {
        val title = contactUsModel.contactUsFinancialService()?.title
        mBottomNavigator?.apply {
            setTitle(title)
            displayToolbar()
            showBackNavigationIcon(true)
        }

        if (activity is MyAccountActivity)
            (activity as? MyAccountActivity)?.setToolbarTitle(title)
    }
}